turbopack/crates/turbopack-ecmascript/src/simple_tree_shake/mod.rs

```diff
@@ -0,0 +1,101 @@
+//! Intermediate tree shaking that uses global information but not good as the full tree shaking.
+
+use anyhow::{Context, Result};
+use rustc_hash::FxHashMap;
+use turbo_rcstr::RcStr;
+use turbo_tasks::{ResolvedVc, Vc};
+use turbopack_core::{
+    module_graph::{ModuleGraph, SingleModuleGraph},
+    resolve::Export,
+};
+
+use crate::chunk::EcmascriptChunkPlaceable;
+
+#[turbo_tasks::function]
+pub async fn is_export_used(
+    graph: ResolvedVc<ModuleGraph>,
+    module: ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>,
+    export_name: RcStr,
+) -> Result<Vc<bool>> {
+    let export_usage_info = compute_export_usage_info(graph)
+        .resolve_strongly_consistent()
+        .await?;
+
+    let export_usage_info = export_usage_info.await?;
+    let Some(exports) = export_usage_info.used_exports.get(&module) else {
+        return Ok(Vc::cell(true));
+    };
+
+    for export in exports {
+        match export {
+            Export::Named(rc_str) => {
+                if rc_str == &export_name {
+                    return Ok(Vc::cell(true));
+                }
+            }
+            Export::Evaluation => {}
+            Export::All => {
+                return Ok(Vc::cell(true));
+            }
+        }
+    }
+
+    Ok(Vc::cell(false))
+}
+
+#[turbo_tasks::function(operation)]
+pub async fn compute_export_usage_info(
+    graph: ResolvedVc<ModuleGraph>,
+) -> Result<Vc<ExportUsageInfo>> {
+    // Layout segment optimization, we can individually compute the async modules for each graph.
+    let mut result: Vc<ExportUsageInfo> = ExportUsageInfo::default().cell();
+    for g in &graph.await?.graphs {
+        result = compute_export_usage_info_single(**g, result);
+    }
+    Ok(result)
+}
+
+#[turbo_tasks::function]
+pub async fn compute_export_usage_info_single(
+    graph: ResolvedVc<SingleModuleGraph>,
+    parent_export_usage_info: ResolvedVc<ExportUsageInfo>,
+) -> Result<Vc<ExportUsageInfo>> {
+    let parent_export_usage_info = parent_export_usage_info.await?;
+
+    let graph = graph.await?;
+    let mut used_exports = FxHashMap::default();
+
+    // Traverse the module graph
+
+    graph
+        .traverse_edges(|(edge, target)| {
+            if let Some(target_module) =
+                ResolvedVc::try_downcast::<Box<dyn EcmascriptChunkPlaceable>>(target.module)
+            {
+                if let Some((_, ref_data)) = edge {
+                    used_exports
+                        .entry(target_module)
+                        .or_insert_with(Vec::new)
+                        .push(ref_data.export.clone());
+                }
+            }
+
+            turbopack_core::module_graph::GraphTraversalAction::Continue
+        })
+        .context("failed to traverse module graph")?;
+
+    for (k, v) in &parent_export_usage_info.used_exports {
```

Comment:
here we copy the parent_export_usage into the output for this function and accumulate it

Doesn't this mean we keep copying used exports as we go? and thus accumulate N^2 memory?

Reply:
I made it cached per SingleModuleGraph and the merging cached per `ResolvedVc<ModuleGraph>`.

The latter should be cached because it's used frequently in a single production build.
I'm not sure if the former should be cached, though.

Reply:
caching per singlemodulegraph will improve incrementality... but i agree the benefits are ambiguous.